certbot/interfaces.py

```diff
@@ -604,3 +604,204 @@ def renew_deploy(self, lineage, *args, **kwargs):
+class BlockNode(ParserNode):
+    """
+    BlockNode class represents a block of nested configuration directives, comments
+    and other blocks as its children. A BlockNode can have zero or more arguments
+    attached to it.
+
+    Configuration blocks typically consist of one or more child nodes of all possible
+    types. Because of this, the BlockNode class has various discovery and structure
+    management methods.
+    """
+
+    @abc.abstractmethod
+    def add_child_block(self, name, arguments=[], position=-1):
+        """
+        Adds a new BlockNode child node with provided values and marks the callee
+        BlockNode dirty. This is used to add new children to the AST.
+
+        :param str name: The name of the child node to add
+        :param list arguments: list of arguments for the node
+        :param int position: Position in the list of children to add the new child
+            node to. Defaults to -1, which appends the newly created node to the list.
+
+        :returns: BlockNode instance of the created child block
+
+        """
+
+    @abc.abstractmethod
+    def add_child_directive(self, name, arguments=[], position=-1):
+        """
+        Adds a new DirectiveNode child node with provided values and marks the
+        callee BlockNode dirty. This is used to add new children to the AST.
+
+        :param str name: The name of the child node to add
+        :param list arguments: list of arguments for the node
+        :param int position: Position in the list of children to add the new child
+            node to. Defaults to -1, which appends the newly created node to the list.
+
+        :returns: DirectiveNode instance of the created child directive
+
+        """
+
+    @abc.abstractmethod
+    def add_child_comment(self, name, arguments=[], position=-1):
+        """
+        Adds a new CommentNode child node with provided values and marks the
+        callee BlockNode dirty. This is used to add new children to the AST.
+
+        :param str name: The name of the child node to add
+        :param list arguments: list of arguments for the node
+        :param int position: Position in the list of children to add the new child
+            node to. Defaults to -1, which appends the newly created node to the list.
+
+        :returns: CommentNode instance of the created child comment
+
+        """
+
+    @abc.abstractmethod
+    def find_blocks(self, name, arg=None):
+        """
+        Find a configuration block with name and (optionally) an argument.
+        This method walks the child tree of ParserNodes under the instance it was
+        called from. This way it is possible to search for the whole configuration
+        tree, when starting from root node or to do a partial search when starting
+        from a specified branch.
+
+        :param str name: The name of the directive to search for
+        :param str arg: Return only directives that have argument with provided value
```

Comment:
Should the `arg` parameter here and in `find_directives` below be a string or a list? Will we ever want to specify more than one argument to search for?

Reply:
I can't think of a case where we would in Apache land. But to make this more generic, a list works better here.

Reply:
From the interface design perspective however, I might be leaning towards removing the arg argument completely, and directing users to do post processing for the directives in order to find the correct ones. This would give more freedom for the user. For example:

```
for vh in rootnode.find_blocks("VirtualHost"):
    if "*:80" in vh.arguments():
        do_a_thing()
```

Reply:
SGTM! We can always add this functionality back if the need/desire arises.